_pythonSummerCourse/04-02-class.md

```diff
@@ -1,21 +1,140 @@
 
-Nội dung kiến thức
-- Khởi tạo class trong Python
-- Các thuộc tính, phương thức của một class
-- `@classmethod`, `@staticmethod`, `@property` decorators
+      def sit(self):
+          print(f"{self.name} is now sitting.")
+  ```
+- Ở ví dụ trên, mỗi người được tạo ra từ lớp Person sẽ được lưu trữ tên (name), tuổi (age) và có hành động ngồi (sit()).
 
-Tài liệu tham khảo
+## 2. Các thuộc tính, phương thức của một class
+### 2.1 Thuộc tính trong class
+- Dựa vào ví dụ trên, các biến `name` và `age` được gọi là thuộc tính.
+- Để truy cập các thuộc tính
+  ```py
+  myFriend = Person('John', 22)
+  print(myFriend.name)  # John
+  ```
+- Một số thuộc tính mặc định (built-in) như `__doc__`, `__class__`, `__name__`, `__dict__`, `__module__`, ...
+  - Ví dụ `__doc__` trong class và trong function cũng có.
+    ```py
+    print(Person.__doc__)  # Initialize a simple class.
+    ```
+  - Ví dụ `__class__`
+    ```py
+    print(myFriend.__class__)  # <class '__main__.Person'>
+    ```
+  - Ví dụ `__name__`
+    ```py
+    print(Person.__name__)  # Person
+    ```
+
+### 2.2. Phương thức trong class
+- Tại ví dụ ở phần 1 (Class trong Python), `sit()` được gọi là 1 phương thức trong class.
+- Để gọi phương thức trong class
+```py
+myFriend.sit()  # John is now sitting.
+```
+
+## 3. `@classmethod`, `@staticmethod`, `@property` decorators
+### 3.1. `@classmethod`
+- Dùng `@classmethod` để đánh dấu cho một class method.
+- Thay vì sử dụng tham số `self` thì class method dùng tham số `cls`.
+  ```py
+  from datetime import date
+
+
+  class Person:
+      """
+      Initialize a simple class.
+      """
+      def __init__(self, name, age):
+          """Initialize name and age attributes."""
+          self.name = name
+          self._age = age
+
+      # a class method to create a Person object by birth year.
+      @classmethod
+      def birthYear(cls, name, year):
+          return cls(name, date.today().year - year)
+
+  personA = Person('John', 22)
+  personB = Person.birthYear('John', 2000)
+
+  print(personA.age)  # 22
+  print(personB.age)  # 22
+  ```
+
+### 3.2. `@staticmethod`
+- Dùng `@staticmethod` để đánh đấu cho một static method.
+- Không dùng tham số `self` và `cls`.
+  ```py
+  class Person:
+      """
+        Initialize a simple class.
+      """
+      def __init__(self, name, age):
+          """Initialize name and age attributes."""
+          self.name = name
+          self._age = age
+
+      # a static method to check if a Person is retirement or not.
+      @staticmethod
+      def isRetirement(age):
+          return age > 65
+
+  print(Person.isRetirement(68))  # True
+  ```
+
+### 3.3 `@property`
+- Trong Python property() được sử dụng như một decorators, có thể sử dụng `@property` để đánh dấu.
+```py
+class Person:
+    """
+        Initialize a simple class.
+    """
+    def __init__(self, name, age):
+        """Initialize name and age attributes."""
+        self.name = name
+        self._age = age
```

Comment:
Một số yêu cầu về trình bày:
- Cần trình bày theo hướng như sau: class Person có đúng hai thuộc tính `name` và `age` (**không có gạch dưới**).
- Khi giới thiệu `property` thì chọn một property với ý nghĩa **hoàn toàn khác**. Không sử dụng cách trình bày thuộc tính là `_age` còn property là `age`.
- Sau khi giới thiệu xong rồi thì trình bày một thuộc tính hoàn toàn khác với kiểu viết _foo, foo, setter, deleter.

Công việc cần thực hiện
- Sửa lại tất cả code theo yêu cầu ở trên.
- Không viết lại class `Person` ở mọi chỗ, cũng như không chỉ viết duy nhất một phiên bản full của class `Person` ngay từ đầu. Cần chọn cách trình bày ở giữa hai thái cực này.